db.js

```diff
@@ -147,53 +158,57 @@ exports.init = function (sbot, config) {
     getHelper(id, false, cb)
   }
 
-  function add(msg, cb) {
+  function getMsgByOffset(offset, cb) {
+    log.get(offset, (err, buf) => {
+      if (err) return cb(err)
+      cb(null, bipf.decode(buf, 0))
+    })
+  }
+
+  function add(msgValue, cb) {
     const guard = guardAgainstDuplicateLogs('add()')
     if (guard) return cb(guard)
 
     onceWhen(
       stateFeedsReady,
       (ready) => ready === true,
       () => {
-        try {
-          state = validate.append(state, hmac_key, msg)
-          if (state.error) return cb(state.error)
-          const kv = state.queue[state.queue.length - 1]
-          log.add(kv.key, kv.value, (err, data) => {
+        const latestMsgValue = state[msgValue.author]
+          ? state[msgValue.author].value
+          : null
+        validate2.validateSingle(null, msgValue, latestMsgValue, (err) => {
+          if (err) return cb(err)
+          const kvt = validate.toKeyValueTimestamp(msgValue) // TODO validate2
+          updateState(kvt)
+          log.add(kvt.key, kvt.value, (err, data) => {
+            if (err) return cb(err)
             post.set(data)
-            cb(err, data)
+            cb(null, data)
           })
-        } catch (ex) {
-          return cb(ex)
-        }
+        })
       }
     )
   }
 
-  function addOOO(msg, cb) {
+  function addOOO(msgValue, cb) {
     const guard = guardAgainstDuplicateLogs('addOOO()')
     if (guard) return cb(guard)
 
-    try {
-      let oooState = validate.initial()
-      validate.appendOOO(oooState, hmac_key, msg)
-
-      if (oooState.error) return cb(oooState.error)
-
-      const kv = oooState.queue[oooState.queue.length - 1]
-      get(kv.key, (err, data) => {
-        if (data) cb(null, data)
-        else
-          log.add(kv.key, kv.value, (err, data) => {
-            post.set(data)
-            cb(err, data)
-          })
+    validate2.validateOOOBatch(null, [msgValue], (err) => {
+      if (err) return cb(err)
+      const kvt = validate.toKeyValueTimestamp(msgValue)
+      get(kvt.key, (err, data) => {
+        if (data) return cb(null, data)
+        log.add(kvt.key, kvt.value, (err, data) => {
+          if (err) return cb(err)
+          post.set(data)
+          cb(null, data)
+        })
       })
-    } catch (ex) {
-      return cb(ex)
-    }
+    })
   }
 
+  // FIXME:
```

Comment:
Don't bother with this :-)

Where I'm currently at is I want to remove that function and instead add a addOOOBatch (together with addBatch (https://github.com/ssb-ngi-pointer/ssb-db2/pull/228)).